File: code/com/leetcode/solutions/NO_0008_字符串转换整数/Solution.java
package com.leetcode.solutions.NO_0008_字符串转换整数;

/**
 * Created by gang.xu01@hand-china.com on 2019/1/6
 */
public class Solution {

    /**
     * 解法：蛮无聊的题，使用asc码判断是否符合规则，最后判断边界问题即可
     *
     * @param str
     * @return
     */
    public int myAtoi(String str) {
        if (str == null || str.trim().equals("")) return 0;
        str = str.trim();
        StringBuilder sb = new StringBuilder();
        char c0 = str.charAt(0);
        if (c0 == 43 || c0 == 45 ||(c0 >= 48 && c0 <=57)) sb.append(c0);
        else return 0;
        for (int i = 1; i < str.length(); i++) {
            char c = str.charAt(i);
            if (c >= 48 && c <= 57) sb.append(c);
            else break;
        }
        try {
            return sb.length() == 1  && (c0 == 43 || c0 == 45) ? 0 :
                    Integer.parseInt(sb.toString());
        } catch (NumberFormatException e) {
            return c0 == 45 ? Integer.MIN_VALUE : Integer.MAX_VALUE;
        }
    }

    public static void main(String[] args) {
        System.out.println(new Solution().myAtoi("3.1"));
    }

}

File: code/com/leetcode/solutions/NO_0008_字符串转换整数/Solution.cs
using System;

public class Solution
{

    /*
   * str:指定字符串
   * 思路：遍历字符串进行判断，最后转换为整型时注意边界问题。
   * 
  */
    public int MyAtoi(string str)
    {
        str = str.Trim();
        if (str.Length == 0) { return 0; }
        int re = 0;
        bool f = true;
        char f0 = str[0];
        try
        {
            if (!f0.ToString().Equals("+") && !f0.ToString().Equals("-")) { int b = int.Parse(f0.ToString()); }

        }
        catch (Exception)
        {
            f = false;
        }
        if (!f) return 0;
        StringBuilder res = new StringBuilder();

        int index = 0;
        foreach (char v in str)
        {
            bool thischar = true;
            if (index == 0 && (v.ToString().Equals("+") || v.ToString().Equals("-")))
            {
                if (v.Equals('-')) { res.Append(v); }
                index++;
                continue;
            }
            try
            {
                int reyy = Convert.ToInt32(v.ToString());
                res.Append(v);
            }
            catch (Exception e)
            {
                thischar = false;
            }
            if (!thischar) { break; }
            index++;
        }
        if (res.ToString().Length > 0 && "0".Equals(res.ToString()[0].ToString())) { res = res.Remove(0, 1); }
        try
        {
            re = Convert.ToInt32(res.ToString());
        }
        catch (OverflowException)
        {
            if ("-".Equals(res.ToString()[0].ToString()))
            {
                re = int.MinValue;
            }
            else { re = int.MaxValue; }

        }
        catch (Exception e) { re = 0; }
        return re;
    }
}
